File: flutter/android/src/main/kotlin/com/spruceid/sprucekit_mobile/Oid4vpMdocAdapter.kt
package com.spruceid.sprucekit_mobile

import android.content.Context
import android.util.Log
import com.spruceid.mobile.sdk.KeyManager
import com.spruceid.mobile.sdk.rs.ApprovedResponse180137
import com.spruceid.mobile.sdk.rs.InProgressRequest180137
import com.spruceid.mobile.sdk.rs.Mdoc
import com.spruceid.mobile.sdk.rs.Oid4vp180137
import com.spruceid.mobile.sdk.rs.RequestMatch180137
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * OID4VP mDoc (ISO 18013-7) Pigeon Adapter for Android
 *
 * Handles OpenID for Verifiable Presentation with mDoc credentials
 */
internal class Oid4vpMdocAdapter(
    private val context: Context,
    private val credentialPackAdapter: CredentialPackAdapter
) : Oid4vpMdoc {

    // Session state
    private var handler: Oid4vp180137? = null
    private var request: InProgressRequest180137? = null
    private var matches: List<RequestMatch180137> = emptyList()
    private var keyManager: KeyManager? = null

    override fun initialize(
        credentialPackIds: List<String>,
        callback: (Result<Oid4vpMdocResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                // Get mDoc credentials from packs
                val mdocs = mutableListOf<Mdoc>()
                for (packId in credentialPackIds) {
                    val packCredentials = credentialPackAdapter.getNativeCredentials(packId)
                    for (credential in packCredentials) {
                        val mdoc = credential.asMsoMdoc()
                        if (mdoc != null) {
                            mdocs.add(mdoc)
                        }
                    }
                }

                if (mdocs.isEmpty()) {
                    callback(Result.success(Oid4vpMdocError("No mDoc credentials found in provided packs")))
                    return@launch
                }

                // Create KeyManager instance
                val keyMgr = KeyManager()

                // Create the handler
                val newHandler = Oid4vp180137(mdocs, keyMgr)

                synchronized(this@Oid4vpMdocAdapter) {
                    this@Oid4vpMdocAdapter.handler = newHandler
                    this@Oid4vpMdocAdapter.keyManager = keyMgr
                }

                callback(Result.success(Oid4vpMdocSuccess("Handler initialized with ${mdocs.size} mDoc(s)")))
            } catch (e: Exception) {
                Log.e("Oid4vpMdocAdapter", "Failed to initialize handler", e)
                callback(Result.success(Oid4vpMdocError(e.localizedMessage ?: "Failed to initialize handler")))
            }
        }
    }

    override fun processRequest(
        url: String,
        callback: (Result<ProcessRequestResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val currentHandler = synchronized(this@Oid4vpMdocAdapter) { handler }
                if (currentHandler == null) {
                    callback(Result.success(ProcessRequestError("Handler not initialized. Call initialize first.")))
                    return@launch
                }

                // Process the request
                val inProgressRequest = currentHandler.processRequest(url)
                val requestMatches = inProgressRequest.matches()

                synchronized(this@Oid4vpMdocAdapter) {
                    this@Oid4vpMdocAdapter.request = inProgressRequest
                    this@Oid4vpMdocAdapter.matches = requestMatches
                }

                if (requestMatches.isEmpty()) {
                    callback(Result.success(ProcessRequestError(
                        "No matching credentials found for this verification request"
                    )))
                    return@launch
                }

                // Convert to Pigeon types
                val matchesData = requestMatches.mapIndexed { index, match ->
                    val fieldsData = match.requestedFields().map { field ->
                        RequestedField180137Data(
                            id = field.id,
                            displayableName = field.displayableName,
                            displayableValue = field.displayableValue,
                            selectivelyDisclosable = field.selectivelyDisclosable,
                            intentToRetain = field.intentToRetain,
                            required = field.required,
                            purpose = field.purpose
                        )
                    }
                    RequestMatch180137Data(
                        index = index.toLong(),
                        credentialId = match.credentialId().toString(),
                        requestedFields = fieldsData
                    )
                }

                val info = Oid4vpMdocRequestInfo(
                    requestedBy = inProgressRequest.requestedBy(),
                    matches = matchesData
                )

                callback(Result.success(ProcessRequestSuccess(info)))
            } catch (e: Exception) {
                Log.e("Oid4vpMdocAdapter", "Failed to process request", e)
                callback(Result.success(ProcessRequestError(e.localizedMessage ?: "Failed to process request")))
            }
        }
    }

    override fun submitResponse(
        matchIndex: Long,
        approvedFieldIds: List<String>,
        callback: (Result<Oid4vpMdocResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val currentRequest = synchronized(this@Oid4vpMdocAdapter) { request }
                val currentMatches = synchronized(this@Oid4vpMdocAdapter) { matches }

                if (currentRequest == null) {
                    callback(Result.success(Oid4vpMdocError("No active request. Call processRequest first.")))
                    return@launch
                }

                if (matchIndex < 0 || matchIndex >= currentMatches.size) {
                    callback(Result.success(Oid4vpMdocError("Invalid match index")))
                    return@launch
                }

                val selectedMatch = currentMatches[matchIndex.toInt()]

                // Create approved response
                val approvedResponse = ApprovedResponse180137(
                    selectedMatch.credentialId(),
                    approvedFieldIds
                )

                // Submit and get redirect URL
                val redirectUrl = currentRequest.respond(approvedResponse)

                callback(Result.success(Oid4vpMdocSuccess(
                    message = "Presentation submitted successfully",
                    redirectUrl = redirectUrl
                )))
            } catch (e: Exception) {
                Log.e("Oid4vpMdocAdapter", "Failed to submit response", e)
                callback(Result.success(Oid4vpMdocError(e.localizedMessage ?: "Failed to submit response")))
            }
        }
    }

    override fun cancel() {
        synchronized(this) {
            handler = null
            request = null
            matches = emptyList()
            keyManager = null
        }
    }
}

File: flutter/ios/Classes/Oid4vpMdocAdapter.swift
import Foundation
import SpruceIDMobileSdk
import SpruceIDMobileSdkRs

/// OID4VP mDoc (ISO 18013-7) Pigeon Adapter for iOS
///
/// Handles OpenID for Verifiable Presentation with mDoc credentials
class Oid4vpMdocAdapter: Oid4vpMdoc {

    private let credentialPackAdapter: CredentialPackAdapter
    private let lock = NSLock()

    // Session state
    private var handler: Oid4vp180137?
    private var request: InProgressRequest180137?
    private var matches: [RequestMatch180137] = []
    private var keyManager: KeyManager?

    init(credentialPackAdapter: CredentialPackAdapter) {
        self.credentialPackAdapter = credentialPackAdapter
    }

    func initialize(
        credentialPackIds: [String],
        completion: @escaping (Result<Oid4vpMdocResult, any Error>) -> Void
    ) {
        Task {
            do {
                // Get mDoc credentials from packs
                var mdocs: [Mdoc] = []
                for packId in credentialPackIds {
                    let packCredentials = credentialPackAdapter.getNativeCredentials(packId: packId)
                    for credential in packCredentials {
                        if let mdoc = credential.asMsoMdoc() {
                            mdocs.append(mdoc)
                        }
                    }
                }

                if mdocs.isEmpty {
                    completion(.success(Oid4vpMdocError(message: "No mDoc credentials found in provided packs")))
                    return
                }

                // Create KeyManager instance
                let keyMgr = KeyManager()

                // Create the handler
                let newHandler = try Oid4vp180137(
                    credentials: mdocs,
                    keystore: keyMgr
                )

                lock.lock()
                self.handler = newHandler
                self.keyManager = keyMgr
                lock.unlock()

                completion(.success(Oid4vpMdocSuccess(message: "Handler initialized with \(mdocs.count) mDoc(s)")))
            } catch {
                completion(.success(Oid4vpMdocError(message: error.localizedDescription)))
            }
        }
    }

    func processRequest(
        url: String,
        completion: @escaping (Result<ProcessRequestResult, any Error>) -> Void
    ) {
        Task {
            do {
                lock.lock()
                guard let handler = self.handler else {
                    lock.unlock()
                    completion(.success(ProcessRequestError(message: "Handler not initialized. Call initialize first.")))
                    return
                }
                lock.unlock()

                // Process the request
                let inProgressRequest = try await handler.processRequest(url: url)
                let requestMatches = inProgressRequest.matches()

                lock.lock()
                self.request = inProgressRequest
                self.matches = requestMatches
                lock.unlock()

                if requestMatches.isEmpty {
                    completion(.success(ProcessRequestError(
                        message: "No matching credentials found for this verification request"
                    )))
                    return
                }

                // Convert to Pigeon types
                let matchesData = requestMatches.enumerated().map { (index, match) -> RequestMatch180137Data in
                    let fieldsData = match.requestedFields().map { field -> RequestedField180137Data in
                        RequestedField180137Data(
                            id: field.id,
                            displayableName: field.displayableName,
                            displayableValue: field.displayableValue,
                            selectivelyDisclosable: field.selectivelyDisclosable,
                            intentToRetain: field.intentToRetain,
                            required: field.required,
                            purpose: field.purpose
                        )
                    }
                    return RequestMatch180137Data(
                        index: Int64(index),
                        credentialId: match.credentialId(),
                        requestedFields: fieldsData
                    )
                }

                let info = Oid4vpMdocRequestInfo(
                    requestedBy: inProgressRequest.requestedBy(),
                    matches: matchesData
                )

                completion(.success(ProcessRequestSuccess(info: info)))
            } catch {
                completion(.success(ProcessRequestError(message: error.localizedDescription)))
            }
        }
    }

    func submitResponse(
        matchIndex: Int64,
        approvedFieldIds: [String],
        completion: @escaping (Result<Oid4vpMdocResult, any Error>) -> Void
    ) {
        Task {
            do {
                lock.lock()
                guard let request = self.request else {
                    lock.unlock()
                    completion(.success(Oid4vpMdocError(message: "No active request. Call processRequest first.")))
                    return
                }

                guard matchIndex >= 0 && matchIndex < matches.count else {
                    lock.unlock()
                    completion(.success(Oid4vpMdocError(message: "Invalid match index")))
                    return
                }

                let selectedMatch = matches[Int(matchIndex)]
                lock.unlock()

                // Create approved response
                let approvedResponse = ApprovedResponse180137(
                    credentialId: selectedMatch.credentialId(),
                    approvedFields: approvedFieldIds
                )

                // Submit and get redirect URL
                let redirectUrl = try await request.respond(approvedResponse: approvedResponse)

                completion(.success(Oid4vpMdocSuccess(
                    message: "Presentation submitted successfully",
                    redirectUrl: redirectUrl
                )))
            } catch {
                completion(.success(Oid4vpMdocError(message: error.localizedDescription)))
            }
        }
    }

    func cancel() throws {
        lock.lock()
        handler = nil
        request = nil
        matches = []
        keyManager = nil
        lock.unlock()
    }
}
